sound/soc/sof/ipc4-pcm.c

```diff
@@ -37,75 +64,92 @@ static int sof_ipc4_trigger_pipelines(struct snd_soc_component *component,
 {
 	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(component);
 	struct snd_soc_pcm_runtime *rtd = asoc_substream_to_rtd(substream);
-	struct snd_sof_widget *pipeline_widget;
-	struct snd_soc_dapm_widget_list *list;
-	struct snd_soc_dapm_widget *widget;
+	struct snd_sof_pipeline_trigger_info *pipeline_info;
+	struct ipc4_pipeline_set_state_data *data;
+	struct snd_sof_widget *pipe_widget;
 	struct sof_ipc4_pipeline *pipeline;
-	struct snd_sof_widget *swidget;
 	struct snd_sof_pcm *spcm;
-	int ret = 0;
-	int num_widgets;
+	u32 pipeline_count = 0;
+	int i = 0;
+	int ret;
 
 	spcm = snd_sof_find_spcm_dai(component, rtd);
 	if (!spcm)
 		return -EINVAL;
 
-	list = spcm->stream[substream->stream].list;
+	pipeline_info = &spcm->stream[substream->stream].pipeline_info;
+
+	/*
+	 * IPC4 requires pipelines to be triggered in order starting at the sink and
+	 * walking all the way to the source. Count the pipeline instance ID's that are not
+	 * already in the requested state and skip those that shouldn't be triggered here.
+	 */
+	list_for_each_entry(pipe_widget, &pipeline_info->pipeline_list, pcm_trigger_list) {
+		pipeline = pipe_widget->private;
+		if (pipeline->state != state && !pipeline->skip_fe_trigger)
+			pipeline_count++;
+	}
 
-	for_each_dapm_widgets(list, num_widgets, widget) {
-		swidget = widget->dobj.private;
+	/* return if all pipelines are in the requested state already */
+	if (!pipeline_count)
+		return 0;
+
+	/* allocate data for the pipeline ID's and the count of pipelines to be triggered */
+	data = kzalloc(sizeof(u32) * (pipeline_count + 1), GFP_KERNEL);
+	if (!data)
+		return -ENOMEM;
+
+	/* set the set pipeline data */
+	data->count = pipeline_count;
+	list_for_each_entry(pipe_widget, &pipeline_info->pipeline_list, pcm_trigger_list) {
+		pipeline = pipe_widget->private;
+		if (pipeline->state != state && !pipeline->skip_fe_trigger)
+			data->pipeline_ids[i++] = pipe_widget->instance_id;
```

Comment:
How does this enforce the 'sink to source' order that's described above?

Reply:
when adding the pipelines, they're added source->sink. So traversing the array in reverse ensures the needed order